enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+- Operator will be deployed in the `openshift-cloud-controller-manager-operator` namespace.
+
+Each provider implementation will need to do the following minimal set of actions in order to make out-of-tree implemetation work:
+
+- Use `external` flag in the `kube-controller-manager` pod instead of the previously used provider. Remove the `cloud-config` option from the template.
+- Use `external` flag in the `kubelet` pod. Remove the `cloud-config` option from the template.
+- Provision a `cloud-controller-manager` `DaemonSet` in the `openshift-cloud-controller-manager` namespace.
+
+#### API resource changes
+
+In order to achieve desired cluster configuration with `external` cloud components, we need to add a new field into `Infrastructure` [resource](https://github.com/openshift/api/blob/c2f7aea5d89ee87d7510b50bdb08dde9028a53ef/config/v1/types_infrastructure.go#L11). With the addition, a `status.platformStatus.external` field will determine the architectural change in the cluster configuration. Other components will be able to consume the configuration same way as before, only with a minor change of the resource processing.
+
+```go
+type PlatformStatus struct {
+  ...
+	// External is a value used in infrastucture automaton to determine component-specific
+	// behavior for the PlatformType value. Components depending on this variable will
+	// ignore cloud specific detailes associated with the given platform type,
+	// and will concider the value of the infrastructure to be "external" - essentially
+	// disabling any further evaluation.
+	//
+	// This will allow to delegate task of processing platform specific bits
+	// to another component.
+	// +optional
+  External bool `json:"external"`
+  ...
+```
+
+To ensure a smooth transition from the `in-tree` selection, `library-go` [implementation](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) will include an additional argument for the `cloudProviderObserver` structure, which will allow to preserve initial configuration strategy, until the new `observeExternal` flag will be set to `false`:
+
+```go
+type cloudProviderObserver struct {
+	targetNamespaceName     string
+	cloudProviderNamePath   []string
+	cloudProviderConfigPath []string
+	observeExternal         bool
+}
+```
+
+The `observeExternal` field will determine, if `status.platformStatus.external` field from the `Infrastructure` resource will be taken in account in establishing needed flags for a Kubernetes binary, such as some `cloud-controller-manager`, `kube-controller-manager`, etc. It will make the observer type to ignore content of the `cloud-config` file, and force the provider to be set as `external` if the option is not selected.
+
+#### Kubelet
```

Comment:
Do the various component need to be updated in a specific order, eg `kcm` then `kubelet`, then `kube-apiserver`?  Are we expecting them to converge to a stable state if there is no order?

Reply:
The order for updating the existing component should not change much, we only need to ensure the `CCM` is running before or shortly after the `kubelet`. It would be great to deploy and run the operator at the same time as `kube-controller-manager` operator is starting.